google/cloud/datastore/helpers.py

```diff
@@ -245,33 +247,35 @@ def get_read_options(eventual, transaction_id, read_time=None):
     :type read_time: datetime
     :param read_time: Read data from the specified time (may be null). This feature is in private preview.
 
+    :type new_transaction_options: :class:`google.cloud.datastore_v1.types.TransactionOptions`
+    :param new_transaction_options: Options for a new transaction.
+
     :rtype: :class:`.datastore_pb2.ReadOptions`
     :returns: The read options corresponding to the inputs.
     :raises: :class:`ValueError` if more than one of ``eventual==True``,
-            ``transaction``, and ``read_time`` is specified.
+            ``transaction_id``, ``read_time``, and ``new_transaction_options`` is specified.
     """
-    if transaction_id is None:
-        if eventual:
-            if read_time is not None:
-                raise ValueError("eventual must be False when read_time is specified")
-            else:
-                return datastore_pb2.ReadOptions(
-                    read_consistency=datastore_pb2.ReadOptions.ReadConsistency.EVENTUAL
-                )
-        else:
-            if read_time is None:
-                return datastore_pb2.ReadOptions()
-            else:
-                read_time_pb = timestamp_pb2.Timestamp()
-                read_time_pb.FromDatetime(read_time)
-                return datastore_pb2.ReadOptions(read_time=read_time_pb)
-    else:
-        if eventual:
-            raise ValueError("eventual must be False when in a transaction")
-        elif read_time is not None:
-            raise ValueError("transaction and read_time are mutual exclusive")
-        else:
-            return datastore_pb2.ReadOptions(transaction=transaction_id)
+    is_set = [
+        bool(x) for x in (eventual, transaction_id, read_time, new_transaction_options)
+    ]
+    if sum(is_set) > 1:
+        raise ValueError(
+            "At most one of eventual, transaction, or read_time is allowed."
```

Comment:
Hmm interesting. My changes were based on the existing python implementation, so I haven't looked too much into the details here. But off the top of my head, it seems like:
- specifying an exact read_time snapshot seems to imply strong consistency. Otherwise, what does it mean to read from a specific timestamp?
- it doesn't make sense to associate an rpc with two different transactions, so it makes sense that transaction and new_transaction are incompatible
- using a transaction comes with it's own implicit read_time, so it makes sense that read_time, transaction, and new_transaction would all be mutually exclusive

I'll have to review the product docs around this. I'm curious how other languages handle this

Reply:
Regarding your first bullet point, it looks like one of the benefits from eventual consistency is that the request will run faster as described in https://cloud.google.com/datastore/docs/concepts/structuring_for_strong_consistency#consistency_levels. So if the read time is a distant point in the past then the data is guaranteed to be available if strong or eventual consistency is used. In this case, allowing for eventual consistency with a specific read_time in the request may be advantageous for the user.

To answer your question, I think if the user runs a query with eventual consistency and a `read_time` then it reads the data from datastore's state at the read_time timestamp, but because it is eventually consistent it may run faster and not wait for other concurrent operations to complete. I see your point though as the docs say "Strongly consistent queries are transactionally consistent, meaning the results are based on a single, consistent snapshot of the data."

What do you think?

Reply:
Yeah, that's a good point. I'm curious if Kristen or Baha have any insights? Do other languages allow multiple options?

Otherwise I'll have to do some more research

Reply:
Actually I just looked at the ReadOptions proto, and this is [declared as a `oneof`](https://github.com/googleapis/googleapis/blob/0e50601ea3d1f828a90d2ddbd52920fcafd461fd/google/datastore/v1/datastore.proto#L563). So I think that gives us our answer